### ISSUE TITLE
Assertion `!RefExpr.isInvalid() && Context.hasSameType(((Expr*) RefExpr.get())->getType(), ParamType.getUnqualifiedType())' failed

### ISSUE THREAD

**Issue by eschnett** [llvm/llvm-project#20540]
|  |  |
| --- | --- |
| Bugzilla Link | [20166](https://llvm.org/bz20166) |
| Version | 3.4 |
| OS | Linux |
| Attachments | [Backtrace](https://user-images.githubusercontent.com/92601316/143749698-27f85d60-4c37-4749-9bce-9b67970f116a.gz), [Run script](https://user-images.githubusercontent.com/92601316/143749699-f201e9cb-abcc-42c6-8314-671178316fe0.gz) |
| CC | @K-ballo,@urnathan |

## Extended Description 

I receive an internal compiler error with Clang 3.4.2. I attach the backtrace, run script, and preprocessed source code. The first few lines of the backtrace are:

```
clang: /home/eschnett/SIMFACTORY/llvm-3.4.2/src/llvm-3.4.2.src/tools/clang/lib/Sema/SemaTemplate.cpp:5121: clang::ExprResult clang::Sema::BuildExpressionFromDeclTemplateArgument(const clang::TemplateArgument&, clang::QualType, clang::SourceLocation): Assertion `!RefExpr.isInvalid() && Context.hasSameType(((Expr*) RefExpr.get())->getType(), ParamType.getUnqualifiedType())' failed.

clang::Sema::DeduceTemplateArguments(clang::ClassTemplatePartialSpecializationDecl*, clang::TemplateArgumentList const&, clang::sema::TemplateDeductionInfo&) + 788
12 clang           0x0000000000d6cc94 clang::Sema::InstantiateClassTemplateSpecialization(clang::SourceLocation, clang::ClassTemplateSpecializationDecl*, clang::TemplateSpecializationKind, bool) + 676
13 clang           0x0000000000d970c6 clang::Sema::RequireCompleteTypeImpl(clang::SourceLocation, clang::QualType, clang::Sema::TypeDiagnoser&) + 1414
```

This occurs on Linux. I built Clang 3.4.2 myself, using GCC 4.9.0. I am also building against GCC 4.9.0's libstdc++ when the error occurs.

**Comment by AaronBallman** [comment]
I downloaded the 49MB of preprocessed source code and could not reproduce the crash, so closing as worksforme. If you're still seeing a crash, please reopen the issue with more steps to reproduce (and hopefully a smaller reproducer).

**Comment by Endilll** [comment]
Fixed in Clang 3.5: https://cexplorer.testlabs.pro/z/K14doq
Reduced by C-Reduce:
```
template <typename b, b>
struct bw
#define by(d) bw<decltype(&d), &d>
#define bz(a, d) by(::d)
    template <typename ca, typename cb, c(ca::*b)>
    struct bw<cb(ca::*), b>
#define cc(a, cd) bz(, cd) c
    struct { void ce(; cc(, ce)
```
